plugin/pkg/utils/namespacedcloudprofile.go

```diff
@@ -0,0 +1,87 @@
+
+package utils
+
+import (
+	"errors"
+	"fmt"
+
+	"k8s.io/apimachinery/pkg/api/equality"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+
+	"github.com/gardener/gardener/pkg/apis/core"
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	"github.com/gardener/gardener/pkg/apis/core/v1beta1/constants"
+	gardencorev1beta1listers "github.com/gardener/gardener/pkg/client/core/listers/core/v1beta1"
+	"github.com/gardener/gardener/pkg/utils/gardener"
+)
+
+// GetCloudProfile determines whether a given CloudProfile (plus namespace) is a valid NamespacedCloudProfile or a CloudProfile and returns the appropriate object
+func GetCloudProfile(cloudProfileLister gardencorev1beta1listers.CloudProfileLister, NamespacedCloudProfileLister gardencorev1beta1listers.NamespacedCloudProfileLister, cp *core.CloudProfileReference, name *string, namespace string) (*gardencorev1beta1.CloudProfile, error) {
+	cloudProfileReference := gardener.BuildCloudProfileReference(name, cp)
+
+	if cloudProfileReference.Kind == constants.CloudProfileReferenceKindNamespacedCloudProfile {
+		var ncpErr error
+		namespacedCloudProfile, ncpErr := NamespacedCloudProfileLister.NamespacedCloudProfiles(namespace).Get(cloudProfileReference.Name)
+		if ncpErr == nil {
+			return &gardencorev1beta1.CloudProfile{Spec: namespacedCloudProfile.Status.CloudProfileSpec}, nil
+		}
+		if !apierrors.IsNotFound(ncpErr) {
+			return nil, ncpErr
+		}
+	}
+
+	cloudProfile, err := cloudProfileLister.Get(cloudProfileReference.Name)
+	if err == nil {
+		return cloudProfile, nil
+	}
+	if !apierrors.IsNotFound(err) {
+		return nil, err
+	}
+	return nil, fmt.Errorf("could not get cloud profile: %+v", err.Error())
+}
+
+// ValidateCloudProfileChanges validates that CloudProfiles referenced are all the same CloudProfile root node or children thereof (NamespacedCloudProfiles with the root node as parent)
+func ValidateCloudProfileChanges(cloudProfileLister gardencorev1beta1listers.CloudProfileLister, namespacedCloudProfileLister gardencorev1beta1listers.NamespacedCloudProfileLister, newShootSpec, oldShootSpec core.ShootSpec, namespace string) error {
+	oldCloudProfileReference := gardener.BuildCloudProfileReference(oldShootSpec.CloudProfileName, oldShootSpec.CloudProfile)
+	if oldCloudProfileReference == nil {
+		return nil
+	}
+	newCloudProfileReference := gardener.BuildCloudProfileReference(newShootSpec.CloudProfileName, newShootSpec.CloudProfile)
+	if equality.Semantic.DeepEqual(oldCloudProfileReference, newCloudProfileReference) {
+		return nil
+	}
+
+	oldCloudProfileRoot, err := GetRootCloudProfile(cloudProfileLister, namespacedCloudProfileLister, oldCloudProfileReference, namespace)
+	if err != nil {
+		return err
+	}
+	newCloudProfileRoot, err := GetRootCloudProfile(cloudProfileLister, namespacedCloudProfileLister, newCloudProfileReference, namespace)
+	if err != nil {
+		return err
+	}
+	if equality.Semantic.DeepEqual(oldCloudProfileRoot, newCloudProfileRoot) {
```

Comment:
Only checking the root CloudProfile here might not be sufficient. Suppose a shoot references a `NamespacedCloudProfile` and uses a machine type, that is only available in the said `NamespacedCloudProfile`. 

Changing the reference for this shoot must not be allowed.

Such a change is only possible if you go down the inheritance.

Reply:
exactly, changed accordingly (see also [other comment](https://github.com/gardener/gardener/pull/10093#discussion_r1679278562))